fixed error in `LibGit2.merge!` when a branch does not have a remote

wildart opened the pull request:
and made exceptions more informative.

wildart committed:
fixed merge error when branch does not have remote

fixed error in an exception message creation when exception isn't GitError
(Found it while looked at #14616)

testing LibGit2.merge! exceptions

removed changes in original message

fixed #13991, rethrowing uknown errors

defer exceptions until update end

moved statement to avoid contamination of backtrace

tkelman commented on base/pkg/entry.jl:
```diff
@@ -376,7 +376,11 @@ function update(branch::AbstractString)
+                        if isa(err, LibGit2.GitError)
+                            print("Package $pkg cannot be updated. Error: $(err.msg) Class: $(err.class)")
+                        else
+                            show(err)
```
this should potentially be a rethrow

wildart replied:
We've already been through it. Rethrowing exception would terminate update process, and the rest of the packages would never be updated until this error is resolved. We do not want to do that, because there are many situation when update of package can fail, i.e. you are in a different branch of the package.

tkelman replied:
Yes, multiple people other than me have said you're catching too many errors here. The way in which the LibGit2 module is coded is bug-prone and needs to be fixed.

wildart replied:
All errors are reported on screen. Is there any problem with that? Why should the update process be stopped because one of the packages cannot be updated?

tkelman replied:
Swallowing errors leads to mystery situations like https://github.com/JuliaLang/julia/issues/13991 - if an exception happened, let's actually have it reported since something is probably broken. Allow through specific things we know are okay, not everything otherwise you end up having hidden typos again.

aviks replied:
>Allow through specific things we know are okay, not everything otherwise you end up having hidden typos again

For what it's worth, while that should be true of library code, I disagree in this situation. This is user interface code, and it should not rethrow, notwithstanding #13991. While that makes debugging difficult, the alternative, we saw in the early days of Pkg1, was  having to delete the entire .julia folder everytime something went wrong.

wildart replied:
As usual truth somewhere in between. I'm going to print exceptions which are clearly come from the package manager functionality and we aware of, such as `PkgError`, `GitError`. The rest will be rethrown.

tkelman replied:
I wouldn't mind deferring a throw until the operation is complete, but just printing errors means no one will ever report them. The libgit2 version of Pkg is still relatively new and undertested, and needs people to find and report issues.

wildart commented:
This approach clashes with a current design. LibGit2 module functions always throw exceptions for any libgit2 error regardless of the severity of the error - some of them are real problems other just minor inconveniences even design features (e.g. supporting protocols). The error can come from libgit2 or LibGit2 wrapper code, we do not want to differentiate at this level (the `GitError` exception is thrown, if it's processed one). So on the level of the package manager we try to not disrupt user experience with any exception which comes as `GitError`. Why?

Example: While I develop package, I created a new branch without any remote. I start update process and it will try to fetch and merge new changes from the remote master branch of the package. Clearly merge will fail for various reasons: dirty repo, different branch, unattached head and so on. So, do we stop update process at this point by throwing exception because libgit2 will report error?  Or do we inform user about a problem and continue update other packages? If process is interrupted until error is fixed (clearly because of exception), I have to finish my development for the package and commit every change, change to master and start update, then change to dev branch after the completion to continue my work.

Sure we want to catch all errors in Pkg, but do we need to create conditions so Pkg become unusable? BTW, error reporting is only in the `Pkg.update` function when there is a loop. The rest of the functions continue to throw exceptions.

tkelman commented:
We should defer the exception until any other background tasks are complete, and still throw it. The task the user asked for still failed and we should inform them of why in a way they will notice, not an easily-missed warning that will scroll by quickly.

wildart commented:
Sure we can defer. but what if there are many exceptions which one to throw? Is there a way to show a stack trace? I'd be happy to to do that.

ivarne commented on base/pkg/entry.jl:
```diff
@@ -383,8 +384,9 @@ function update(branch::AbstractString)
-                            rethrow(err)
+                            print_with_color(:red, "Package $pkg cannot be updated. Error:$err")
                         end
+                        push!(defered_errors, CapturedException(err, catch_backtrace()))
```
I'd move this line right after ``catch err``, because `catch_backtrace()` might be overwritten by calling other code (not sure if that is the case here.)

tkelman commented on base/pkg/cache.jl:
```diff
@@ -43,8 +43,15 @@ function prefetch(pkg::AbstractString, url::AbstractString, sha1s::Vector)
+            elseif isa(err, InterruptException)
+                "Package `$pkg` prefetching was interrupted."
+            else
+                "Uknown error: $err"
```
Unknown

wildart committed:
fixed similar error in `rebase!`, fixed typos

wildart committed:
Separate internal API and Pkg UI commands by using `PkgError`
`PkgError` becomes an information error that would keep underlying exceptions
`CapturedException` should have supertype `Exception`

tkelman commented on base/pkg.jl:
```diff
@@ -11,7 +11,9 @@ const META_BRANCH = "metadata-v2"
     msg::AbstractString
+    ex::Nullable{Exception}
 end
+PkgError(msg::AbstractString) = PkgError(msg, nothing)
```
doesn't seem right to initialize a `Nullable` field with `nothing`

tkelman commented on base/pkg/entry.jl:
```diff
@@ -328,48 +328,48 @@ end
         try
             Cache.prefetch(pkg, Read.url(pkg), [a.sha1 for (v,a)=avail[pkg]])
         catch err
             warn("Package $pkg: unable to update cache\n$(err.msg)")
```
not all types of exceptions `err` are going to have a `msg` field

tkelman commented on base/task.jl:
```diff
@@ -3,7 +3,7 @@
 
 # Container for a captured exception and its backtrace. Can be serialized.
-type CapturedException
+type CapturedException <: Exception
```
@amitmurthy did you have a specific reason for making this not a subtype of Exception?

amitmurthy replied:
It was designed to be a container for an exception and its backtrace that can be safely serialized. Since it was not the exception itself, I guess I didn't make it a subtype of Exception.

wildart replied:
Is is going to be a problem to be a subtype of Exception?

amitmurthy replied:
I can't think of any.

wildart committed:
Put off `Cache.prefetch` exceptions in `Pkg.update`

wildart committed:
enable `fetch` exceptions, nice `PkgError` output

wildart commented:
Ready to merge.

tkelman commented:
lgtm at 141ba0fa273290b59da4c32ff7e338ea53d04093 but would like to leave it open a bit longer for others to review

wildart committed:
fix possible undefined behavior (ref pull/14146)

wildart commented:
I need to merge this to proceed with cached credentials update.

tkelman commented on base/libgit2.jl:
```diff
@@ -388,9 +386,13 @@ function merge!(repo::GitRepo;
                 with(upstream(head_ref)) do tr_brn_ref
+                    if tr_brn_ref === nothing
+                        throw(GitError(Error.Merge, Error.ERROR,
+                                   "There is no tracking information for the current branch."))
```
indentation is all over the place here. stick to consistent 4 spaces

tkelman commented on base/pkg/entry.jl:
```diff
@@ -383,8 +391,9 @@ function update(branch::AbstractString)
-                        show(err)
-                        print('\n')
+                        push!(deferred_errors,
+                                PkgError("Package $pkg cannot be updated.",
```
here too

wildart committed:
fixed indentations
